[ticket/11150] Install extensions through composer

Nicofuma opened the pull request:
https://tracker.phpbb.com/browse/PHPBB3-11150

TODO:
- [x] A few language keys are missing
- [x] Add debug settings (to display more or less info in the composer output)
- [x] Titania (https://github.com/phpbb/customisation-db/pull/149)
- [x] Display a message instead of the catalog if it is empty? (notice, blue)

Further PR:
 - [ ] Handles board updates (disable, update, enable exts/checks if exts are still compatible)
 - [ ] Deal with exts added manually (checks compatibility)

With titania only: https://blackfire.io/profiles/d8e6c3ff-d784-43cd-a080-809b03fa0d91/graph
With packagist and titania: https://blackfire.io/profiles/8d07d01b-9c7b-4518-9ae7-b75645d25634/graph

PHPBB3-11150

VSEphpbb commented on phpBB/language/en/cli.php:
```diff
@@ -110,6 +110,18 @@
+	'CLI_DESCRIPTION_EXTENSION_REMOVE_OPTION_PURGE'		=> 'Purge the extensions before removing them',
+	'CLI_DESCRIPTION_EXTENSION_REMOVE_ARGUMENT'			=> 'Extensions to remove',
+	'CLI_DESCRIPTION_EXTENSION_UPDATE'					=> 'Updates extensions',
+	'CLI_DESCRIPTION_EXTENSION_UPDATE_ARGUMENT'			=> 'Extensions to update',
```
I'm still not sure about how to call/describe MANAGE. But for the others:
```
'CLI_DESCRIPTION_EXTENSION_INSTALL'					=> 'Install the specified extension(s).',
'CLI_DESCRIPTION_EXTENSION_INSTALL_OPTION_ENABLE'	=> 'Enable extension(s) after installation',
'CLI_DESCRIPTION_EXTENSION_INSTALL_ARGUMENT'		=> 'Extension(s) to install, e.g.: vendor/package',
'CLI_DESCRIPTION_EXTENSION_LIST_AVAILABLE'			=> 'List extensions available for installation.',
'CLI_DESCRIPTION_EXTENSION_REMOVE'					=> 'Remove the specified extension(s).',
'CLI_DESCRIPTION_EXTENSION_REMOVE_OPTION_PURGE'		=> 'Purge extension(s) when removing them',
'CLI_DESCRIPTION_EXTENSION_REMOVE_ARGUMENT'			=> 'Extension(s) to remove, e.g.: vendor/package',
'CLI_DESCRIPTION_EXTENSION_UPDATE'					=> 'Update the specified extension(s).',
'CLI_DESCRIPTION_EXTENSION_UPDATE_ARGUMENT'			=> 'Extension(s) to update, e.g.: vendor/package',
```

Nicofuma commented on phpBB/phpbb/db/migration/data/v320/extensions_composer.php:
```diff
@@ -0,0 +1,37 @@
+	public function update_data()
+	{
+		return array(
+			array('config.add', array('exts_composer_repositories', serialize([]))),
```
add titania once https://github.com/phpbb/customisation-db/pull/149 is merged

marc1706 replied:
Why are you using serialize and not the json format?

Nicofuma commented on phpBB/phpbb/db/migration/data/v320/extensions_composer.php:
```diff
@@ -0,0 +1,37 @@
+	{
+		return array(
+			array('config.add', array('exts_composer_repositories', serialize([]))),
+			array('config.add', array('exts_composer_packagist', true)),
```
switch to false once https://github.com/phpbb/customisation-db/pull/149 is merged

Nicofuma commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,677 @@
+	/**
+	 * @var array Repositories to look packages from
+	 */
+	protected $repositories = [];
```
add titania once https://github.com/phpbb/customisation-db/pull/149 is merged?

marc1706 replied:
probably a good idea, yeah

Nicofuma replied:
done in the migration script

VSEphpbb commented on phpBB/language/en/acp/common.php:
```diff
@@ -88,6 +88,11 @@
 	'ACP_EXTENSION_GROUPS'		=> 'Manage attachment extension groups',
 	'ACP_EXTENSION_MANAGEMENT'	=> 'Extension management',
 	'ACP_EXTENSIONS'			=> 'Manage extensions',
+	'ACP_EXTENSIONS_GALLERY'	=> 'Extensions gallery',
```
We should find a replacement for gallery. maybe Catalogue

rxu replied:
Extensions market ;)

VSEphpbb replied:
that implies $$
My ideas:
Catalog
Depot
Database
Collection
Exhibit
Browse Extensions

VSEphpbb commented on phpBB/phpbb/composer/io/translate_composer_trait.php:
```diff
@@ -0,0 +1,237 @@
+	public function write($messages, $newline = true)
+	{
+		$messages = (array) $messages;
+		$translated_messages = [];
```
Is there a reason not to use traditional array syntax?

Nicofuma replied:
short syntax is better ;)

VSEphpbb replied:
PHP 5.4.0 only. Is that officially the new min. req. yet? I still see 5.3.9 as our min. req.

Nicofuma replied:
It's gonna change soon

VSEphpbb replied:
Well the linters don't like it. ;)

VSEphpbb commented on phpBB/language/en/acp/common.php:
```diff
@@ -226,6 +231,17 @@
+
+	'COMPOSER_UNSUPPORTED_OPERATION'	=> 'Operation unsupported for the package type “%s”.',
+	'COMPOSER_UPDATING_DEPENDENCIES'	=> 'Updating packages',
+	'COMPOSER_LOADING_REPOSITORIES'		=> 'Loading remote packages information',
```
Loading remote repositories with package information

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+
+	'EXTENSION_ALREADY_INSTALLED'				=> 'The “%s” extension has already been installed.',
+	'EXTENSION_ALREADY_INSTALLED_MANUALLY'		=> 'The “%s” extension has already been installed manually.',
+	'EXTENSION_ALREADY_MANAGED'					=> 'The “%s” extension has already been installed manually.',
```
Is this right? Seems a duplicate of above line.

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+	'EXTENSION_ALREADY_INSTALLED'				=> 'The “%s” extension has already been installed.',
+	'EXTENSION_ALREADY_INSTALLED_MANUALLY'		=> 'The “%s” extension has already been installed manually.',
+	'EXTENSION_ALREADY_MANAGED'					=> 'The “%s” extension has already been installed manually.',
+	'EXTENSION_CANNOT_MANAGE_FILESYSTEM_ERROR'	=> 'The “%s” extension cannot be managed because the old files cannot be removed.',
```
The “%s” extension cannot be managed because the existing files could not be removed from the filesystem.

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+	'EXTENSION_CANNOT_MANAGE_INSTALL_ERROR'		=> 'The “%s” extension cannot be installed. The old extension have been restored.',
+	'EXTENSION_MANAGED_WITH_CLEAN_ERROR'		=> 'The “%1$s” extension has been installed but an error occurred and the old files have not been removed. You might want to delete the “%2$s” files manually.',
+	'EXTENSION_MANAGED_WITH_ENABLE_ERROR'		=> 'The “%s” extension has been installed but an error occurred when re-enabling it.',
+	'EXTENSION_NOT_INSTALLED'					=> 'The “%s” extension is not installed.',
```
```
'EXTENSION_CANNOT_MANAGE_INSTALL_ERROR'		=> 'The “%s” extension could not be installed. The prior installation of this extension has been restored.',
'EXTENSION_MANAGED_WITH_CLEAN_ERROR'		=> 'The “%1$s” extension has been installed but an error occurred and the old files have not been removed. You might want to delete the “%2$s” files manually.',
EXTENSION_MANAGED_WITH_ENABLE_ERROR'		=> 'The “%s” extension has been installed but an error occurred when enabling it.',
```

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+	'DISABLING_EXTENSIONS'	=> 'Disabling extensions',
+
+	'EXTENSIONS_GALLERY' => 'Extensions Gallery',
+	'EXTENSIONS_GALLERY_EXPLAIN' => 'The Extensions Gallery is a tool in your phpBB Board which allows you to see all the extensions that are available for your board.',
```
Here you can browse all of the extensions available for your phpBB board. Extensions can easily be installed or removed with just a click. Adjust the Settings to allow instant enabling and purging of extensions.

Nicofuma commented:
The minimum-stability flag is now correctly handled and configurable.

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -129,4 +148,30 @@
+	'STABILITY_DEV'		=> 'dev',
+
+	'COMPOSER_MINIMUM_STABILITY'			=> 'Minimum stability',
+	'COMPOSER_MINIMUM_STABILITY_EXPLAIN'	=> 'TODO: why it is dangerous to change that',
```
```html
<strong class="error">Warning:</strong> Always use <samp>stable</samp> versions on a live forum.
Non-stable versions may still be in development and could cause unexpected problems with your
forum and should only be used for development purposes in local or staging environments.
```

VSEphpbb commented:
I wonder if the Repositories, Packagist and Min Stability settings should all be hidden and ignored except in the development environment, as these settings seem "dangerous" to have available for a live forum?

VSEphpbb commented:
If colors in the composer output can be controlled, the version #'s should be changed. Currently it's bright yellow which is unreadable against the grey bg of the ACP.

Nicofuma commented:
@VSEphpbb I changed the color, should be better now. I have also fixed your issue with enable on install/purge on remove

Nicofuma commented:
@naderman Mind have a look for the composer part and the settings?

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -129,4 +148,32 @@
+
+	'COMPOSER_MINIMUM_STABILITY'			=> 'Minimum stability',
+	'COMPOSER_MINIMUM_STABILITY_EXPLAIN'	=> 'Always use <samp>stable</samp> versions on a live forum.
+Non-stable versions may still be in development and could cause unexpected problems with your
```
the line breaks aren't really needed.

Nicofuma replied:
It's easier to read in the language file...  but yes. Fixed

Nicofuma commented:
For the remove action, 2 options:
- Purge by default when the extension is removed
- Display the action only when purge on remove is enabled or the extension is already purged

VSEphpbb commented on phpBB/adm/style/detailled_message_body.html:
```diff
@@ -0,0 +1,15 @@
+	<p>{{ MESSAGE_TEXT }}</p>
+</div>
+
+<fieldset>
```
Wrap this field set inside of a `<div style="overflow-x:scroll">`

Nicofuma replied:
thanks

Nicofuma commented:
The remove action is now displayed only for disabled extensions (and in the catalog), and the extensions are purged when removed by default.

naderman commented:
@Nicofuma Won't get to it before the weekend, but put it on my calendar.

VSEphpbb commented:
In the extension catalog, enabled extensions still have the Remove link shown.

There is also still the possibility to Remove an extension here, before it has been purged. I think the solution is:

Remove only appears (on both pages) for purged extensions (and then we don't need the `exts_composer_purge_on_remove` setting)

This way, Remove will always only be able to occur after an extension is deliberately purged by the user, and will avoid errors in the ext manager like:
> The “phpbb/pages” extension is not valid.
The requested file could not be found: ./../ext/phpbb/pages/composer.json
And will avoid possible deletions of extension data by the users

VSEphpbb commented:
Also, on further thought, I think having "Remove" and "Update" actions in both pages may be too much (clutter, confusion, redundancy). Consider the Extension Catalog just having "Install" action, and "Installed" status. And Leave the "Remove" and "Update" action for the Extensions Manager (where they are more applicable).

paul999 commented on phpBB/adm/style/acp_ext_catalog.html:
```diff
@@ -0,0 +1,135 @@
+
+<h1>{{lang( 'EXTENSIONS_CATALOG') }}</h1>
+
+<p>{{lang( 'EXTENSIONS_CATALOG_EXPLAIN') }}</p>
```
Missing space after {{ (Or on line 11 there is a space to much)

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -324,6 +351,315 @@ function main()
+				$this->page_title = 'ACP_EXTENSIONS_CATALOG';
+				$this->tpl_name = 'acp_ext_catalog';
+
+				$this->template->assign_var('extensions', $extensions);
```
Why not use $template->assign_vars for these assignments?

paul999 commented on phpBB/phpbb/composer/manager_interface.php:
```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * Created by IntelliJ IDEA.
```
Wrong copyright header

naderman commented:
So how practical is this with memory restrictions and max execution time in place on a shared hoster? I wonder if to make this practical we will have to offer a service on phpbb.com that calculates the actual dependencies to install?

VSEphpbb commented:
128MB memory was insufficient ( @Nicofuma made a PR at composer to try to relieve some of that stress)

naderman commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,694 @@
+			->setUpdateWhitelist($whitelist)
+			->setWhitelistDependencies(false)
+			->setIgnorePlatformRequirements(false)
+			->setDumpAutoloader(false)
```
won't we need the autoloader?

Nicofuma replied:
At first I thought no, but actually... we need it for the exts dependencies

naderman commented:
In general the implementation of the Composer part makes sense I think. However currently the locking mechanism is entirely unused and every install/removal/update triggers a full update of all packages?

naderman commented:
@VSEphpbb that PR won't get us remotely close to 128MB, so I don't really see this working out on a shared host to be honest :(

Nicofuma commented:
For the time it's not an issue because the ma_execution_time doesn't take the system and http call (like file_get_contents) in account.

For the memory, currently it's around 130MB, with my PR it's around 85 - 90MB (didn't try with yours)

Finally, I think it is using the lock file

Nicofuma commented:
@phpbb/development-team please review and comment, especially if you don't agree/have some reserve

naderman commented:
@Nicofuma Well the low amount of memory is due to the relatively small size of the repository - no? So if we encourage phpBB authors to make more use of existing libraries, this will become a lot worse?

naderman commented:
Anyway I guess that's good enough for now. Unfortunately my PR isn't quite ready yet, need to work on that some more soon.

Nicofuma commented:
It's due to the low amount of packages yes. But I guess/hope your PR will be ready before titania start serving a 80Mb json file :)

naderman commented:
Well even with my PR you can't expect more than maybe a 10 percent improvement. We're way beyond trivial major performance improvements.

Nicofuma commented:
It may open the door for other improvements :)

VSEphpbb commented on phpBB/adm/style/acp_ext_catalog.html:
```diff
@@ -0,0 +1,132 @@
+<div class="errorbox">
+	{{ lang('EXTENSIONS_COMPOSER_NOT_WRITABLE') }}
+</div>
+{% endif %}I
```
`I` ?

VSEphpbb commented:
My only main concern is regarding the Remove button. I think it should only be displayed next to extensions that have been purged. Right now it still displays next to extensions that have only been disabled, allowing users to accidentally bypass the purge step, leading to improperly uninstalled extensions. You can also eliminate the "Purge extensions when removing" setting.

Also, copying Pagination to the top is unnecessary and has broken the Browse Extensions and Settings links layout badly.

Lastly, it appears in the composer output that slashes have been converted to `$`? For example:
`packagist.org/provider-phpbb$pages.json from cache`

Nicofuma commented:
For the last point that's how packagist works :) for the others I will fix that soon.

prototech commented on phpBB/adm/style/detailled_message_body.html:
```diff
@@ -0,0 +1,17 @@
+<!-- INCLUDE overall_header.html -->
```
`detailed` instead of `detailled` in the file name.

Nicofuma commented:
pushed back to 3.3 has we need more time to test and decide what to do

marc1706 commented:
This is very unfortunate. :cry:

naderman commented:
@Nicofuma Sure we can't just test/fix this up during alpha/beta phase? Would be rather good to finally get this in?

Nicofuma commented:
That's possible yes, I just need time and that's not my priority for now (I'm planning the a2 for the end of the week). But that could be nice indeed.

We can also think about it for 3.2.x if it's not ready in time for 3.2.0 (it doesn't involve any BC break
That's possible yes, I just need time and that's not my priority for now (I'm planning the a2 for the end of the week). But that could be nice indeed. I'll try to rebase/update with titania this week.

Nicofuma commented:
Rebased and updated to support titania.
I have also included 2 Blackfire profiles (in the PR description) showing the resources usage with and without packagist.

It should work as excepted but a lot of wording and some small things are missing

paul999 commented:
If you need any help with it let me know, I might have some spare time this weekend/next weekend, and would be great to get this into 3.2

Nicofuma commented:
Any help is welcome :) You can test it and help with the wording (mostly the settings and the warning when you enable packagist usage). It could be nice to provide a better feedback when one directory or file isn't writable.

VSEphpbb commented:
A lot better response with just Titania and no packagist :)

marc1706 commented:
The permissions of phpBB/ext/index.html shouldn't have been changed. Please change them back to 644.

marc1706 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -324,6 +351,317 @@ function main()
+					)
+				);
+
+				break;
```
break should have the same indentation as the case. You can change phpstorm to properly do that.

marc1706 commented on phpBB/language/en/acp/common.php:
```diff
@@ -226,6 +231,17 @@
+	'COMPOSER_UPDATING_DEPENDENCIES'	=> 'Updating packages',
+	'COMPOSER_LOADING_REPOSITORIES'		=> 'Loading remote repositories with package information',
+	'COMPOSER_ERROR_CONFLICT'			=> 'Your requirements could not be resolved to an installable set of packages.',
+	'COMPOSER_REPOSITORY_UNAVAILABLE'	=> 'An error occurred when fetching the repository %s.',
```
while fetching

marc1706 commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+	'EXTENSION_ALREADY_MANAGED'					=> 'The “%s” extension is already managed.',
+	'EXTENSION_CANNOT_MANAGE_FILESYSTEM_ERROR'	=> 'The “%s” extension cannot be managed because the existing files could not be removed from the filesystem.',
+	'EXTENSION_CANNOT_MANAGE_INSTALL_ERROR'		=> 'The “%s” extension could not be installed. The prior installation of this extension has been restored.',
+	'EXTENSION_MANAGED_WITH_CLEAN_ERROR'		=> 'The “%1$s” extension has been installed but an error occurred and the old files have not been removed. You might want to delete the “%2$s” files manually.',
```
the old files could not be removed.

marc1706 commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+	'EXTENSION_CANNOT_MANAGE_FILESYSTEM_ERROR'	=> 'The “%s” extension cannot be managed because the existing files could not be removed from the filesystem.',
+	'EXTENSION_CANNOT_MANAGE_INSTALL_ERROR'		=> 'The “%s” extension could not be installed. The prior installation of this extension has been restored.',
+	'EXTENSION_MANAGED_WITH_CLEAN_ERROR'		=> 'The “%1$s” extension has been installed but an error occurred and the old files have not been removed. You might want to delete the “%2$s” files manually.',
+	'EXTENSION_MANAGED_WITH_ENABLE_ERROR'		=> 'The “%s” extension has been installed but an error occurred when enabling it.',
```
while enabling it

marc1706 commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+
+	'EXTENSION_ALREADY_INSTALLED'				=> 'The “%s” extension has already been installed.',
+	'EXTENSION_ALREADY_INSTALLED_MANUALLY'		=> 'The “%s” extension has already been installed manually.',
+	'EXTENSION_ALREADY_MANAGED'					=> 'The “%s” extension is already managed.',
```
In what context is this being used?

Nicofuma replied:
manage with composer an ext installed manually (basically: uninstall and reinstall with composer)

marc1706 commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -35,6 +35,22 @@
+	'DISABLING_EXTENSIONS'	=> 'Disabling extensions',
+
+	'EXTENSIONS_CATALOG'			=> 'Extensions Catalog',
+	'EXTENSIONS_CATALOG_EXPLAIN'	=> 'Here you can browse all of the extensions available for your phpBB board. Extensions can easily be installed or removed with just a click. Adjust the Settings to allow instant enabling and purging of extensions.',
```
settings should be lowercase

marc1706 commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,695 @@
+	private $ext_json_file_backup;
+
+	/**
+	 * @param \phpbb\config\config	$config		Config object
```
docblock doesn't match declaration

marc1706 commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,695 @@
+	{
+		if ($config)
+		{
+			$repositories        = json_decode($config['exts_composer_repositories'], true);
```
Use tabs instead of space if the indentation is really wanted

naderman replied:
Another thing I think we should get rid of in coding guidelines, inline indentation really only causes overly complicated diffs later on.

Nicofuma replied:
Maybe we/you can prepare a post stating what we want to change in the coding style (the long term goal) and then announce it?

paul999 commented on phpBB/bin/phpbbcli.php:
```diff
@@ -68,6 +68,11 @@
 
+if (is_file($phpbb_root_path . $config['exts_composer_vendor_dir'] . '/autoload.php'))
+{
+	require_once($phpbb_root_path . $config['exts_composer_vendor_dir'] . '/autoload.php');
```
should we do some sanitizing here on $config['exts_composer_vendor_dir']?

marc1706 replied:
Can this even be changed by a user? If it can't, then this is a non-issue.

paul999 replied:
It is in the database, so in theory it can be changed by the user, either via a database tool or via the restore functionality in the ACP.
Does it actually need to be configurable?

Nicofuma replied:
About sanitization, I think it has to be done on insertion, not on usage. And because we don't provide any interface to do that (only the config:set command which is a tool for advanced user). I think we can assume that the board owner can take its responsibilities.

About making it configurable... I think it's necessary because some installation wouldn't give write access to anyone in some random places but concentrate them in a specific subfolder. We have the same issue for ext/ composer-ext.json and composer-ext.lock. But for them we can't do anything for now. IIRC the composer-ext.json has to be in the root directory otherwise the custom installers (the one installed phpBB ext/styles/lang pack in ext/ styles/ and languages/) will fail. And for ext/ it's related to this composer issue and the core isn't ready yet to put the extensions in a random place (but it could a matter of adding a new config setting)

paul999 replied:
If we are considering that anything from the DB is safe to use, then I guess we are fine.

Nicofuma replied:
I think it's better to ensure that what we put in the DB is validated than having to revalidate anything on usage yes (one reason being that this way the extensions doesn't have to revalidate the data they are using)

paul999 commented on phpBB/phpbb/composer/extension_manager.php:
```diff
@@ -0,0 +1,316 @@
+	public function pre_install(array $packages, IOInterface $io = null)
+	{
+		$installed_manually = array_intersect(array_keys($this->extension_manager->all_available()), array_keys($packages));
+		if (count($installed_manually) !== 0)
```
Sizeof (http://area51.phpbb.com/docs/master/coding-guidelines.html#general --> General function usage)

nickvergessen replied:
!empty()

Nicofuma replied:
honestly I don't like using empty(), it's doing too much thing...

marc1706 replied:
To be honest, we might want to change that to count() anyway in the coding guidelines, especially since sizeof() might be confusing for people coming from other programming languages.

Nicofuma replied:
according to the doc sizeof() is an alias of count() by the way (and not the opposite). I think that's why phpstorm marks sizeof() as deprecated

paul999 replied:
Sure, I know, but the coding guidelines says we should use sizeof. As far as I know we all want to keep ourselfs on the coding guidelines, so as long that those aren't changed we should use sizeof.

Nicofuma replied:
yep

naderman replied:
I agree that we should probably change it to count() we could simply change it to allow either, but prefer count() for new changes for the time being. So if we merge that change to the coding guidelines before this, we can leave the count as is here.

Nicofuma commented:
updated

marc1706 commented on phpBB/phpbb/di/container_builder.php:
```diff
@@ -364,6 +364,12 @@ protected function load_extensions()
 			$ext_container->register('cache.driver', '\\phpbb\\cache\\driver\\dummy');
 			$ext_container->compile();
 
+			$config = $ext_container->get('config');
```
Seems like this is not available in tests which is also why tests currently fail.

marc1706 commented:
I guess this should be taken care of in the installer?
```In order to use the catalog, the following files and directories must be writable: ext/ vendor-ext/ composer-ext.json and composer-ext.json.lock```

marc1706 commented:
Can we maybe try to enable titania's composer page? That way this is actually testable.

Nicofuma commented:
It is already enabled and working :)

VSEphpbb commented:
@Nicofuma Sorting exts needs to be case insensitive

VSEphpbb commented:
Had a problem. An ext was marked as installed manually, but had the "Manage" option, so I clicked it. Everything appeared to go fine, but it installed it wrong (thus broken ext).

It installed `ext/vse/abbc3` to: `ext/vse/abbc3/abbc3`

VSEphpbb commented on phpBB/adm/style/detailed_message_body.html:
```diff
@@ -0,0 +1,17 @@
+</div>
+
+<div style="overflow-x:scroll">
+	<fieldset>
```
To get a console look, add a class to this element with:
```css
background-color: #2a2a2a;
color: #f1f1f1;
font-family: monospace;
```
Also remove the legend below

VSEphpbb commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,695 @@
+
+			usort($available, function($a, $b)
+			{
+				return strcmp($a['display_name'], $b['display_name']);
```
use `strtoupper` on these so the sorting is case insensitive?

VSEphpbb replied:
or just use `strcasecmp`

VSEphpbb commented on phpBB/adm/style/admin.css:
```diff
@@ -2578,3 +2588,10 @@ fieldset.permissions .padding {
 }
+
+.console-output {
+	background-color: #2a2a2a;
```
I'd also add `line-height: 1.5em;`

hanakin replied:
except unitless can't use `em` for line-height

VSEphpbb commented on phpBB/phpbb/composer/io/html_output_formatter.php:
```diff
@@ -0,0 +1,88 @@
+			{
+				$out .= 'color:'.self::$availableForegroundColors[$code].';';
+			}
+			elseif (isset(self::$availableBackgroundColors[$code]))
```
else if

VSEphpbb commented on phpBB/phpbb/composer/io/html_output_formatter.php:
```diff
@@ -0,0 +1,88 @@
+			{
+				$out .= 'background-color:'.self::$availableBackgroundColors[$code].';';
+			}
+			elseif (isset(self::$availableOptions[$code]))
```
else if

Nicofuma commented:
Updated

paul999 commented:
Currently, the installer doesn't check for permissions/existsing of some files, causing notices to happens:
[phpBB Debug] PHP Warning: in file [ROOT]/vendor/composer/composer/src/Composer/Json/JsonFile.php on line 86: file_get_contents(composer-ext.json): failed to open stream: No such file or directory
[phpBB Debug] PHP Warning: in file [ROOT]/vendor/composer/composer/src/Composer/Json/JsonFile.php on line 123: file_put_contents(composer-ext.json): failed to open stream: Permission denied

It does give some nice message to the user, but the notices should not be there. 

After creating the directories, nothing is shown here in the list (I guessed something should show up to install in the catalog?)

paul999 commented on phpBB/adm/style/acp_ext_catalog.html:
```diff
@@ -0,0 +1,132 @@
+{% endif %}
+
+<fieldset class="quick quick-left">
+	<span class="small"><a href="https://www.phpbb.com/go/customise/extensions/3.1" target="_blank">{{ lang('BROWSE_EXTENSIONS_DATABASE') }}</a> &bull; <a href="javascript:phpbb.toggleDisplay('catalog_settings');">{{ lang('SETTINGS') }}</a></span>
```
Should go to 3.2 instead of 3.1

paul999 commented:
Missing lang key:
COMPOSER_CANNOT_INSTALL


It is using the wrong path to store (Note the directory name phpBBstore):
```
Cannot create cache directory /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/cache/repo/http-/, or directory is not writable. Proceeding without cache
Cannot create cache directory /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/cache/repo/https---packagist.org/, or directory is not writable. Proceeding without cache
Cannot create cache directory /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/cache/files/, or directory is not writable. Proceeding without cache
Failed to initialize global composer: Composer could not find the config file: /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/composer.json```

After creating the directory, it gives:
```
COMPOSER_CANNOT_INSTALL

The "http:/packages.json" file could not be downloaded: failed to open stream: No such file or directory

« Back to previous page

Reading composer-ext.json
Loading config file composer-ext.json
Executing command (/home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBB): git describe --exact-match --tags
Executing command (/home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBB): git branch --no-color --no-abbrev -v
Failed to initialize global composer: Composer could not find the config file: /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/composer.json
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
Reading composer-ext.lock
Loading composer repositories with package information
Downloading http:/packages.json
Downloading http:/packages.json
Downloading http:/packages.json
```

Extension I tried was phpbb/boardrules

paul999 commented:
Ok, found the issue. We still need to add titania as composer repo ( https://www.phpbb.com/customise/db/composer/packages-extension.json ), after that setting it to stable again it shows up right versions and a new error :P

```
COMPOSER_CANNOT_INSTALL

The "https://www.phpbb.com/customise/db/composer/packages-extension.json/packages-extension-1.json" file could not be downloaded (HTTP/1.1 404 Not Found)
```
Note the total path (packages-extension.json/packages-extensions-1.json).
```
Reading composer-ext.json
Loading config file composer-ext.json
Executing command (/home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBB): git describe --exact-match --tags
Executing command (/home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBB): git branch --no-color --no-abbrev -v
Failed to initialize global composer: Composer could not find the config file: /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/composer.json
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
Reading composer-ext.lock
Loading composer repositories with package information
Downloading https://www.phpbb.com/customise/db/composer/packages-extension.json
Writing /home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBBstore/composer/cache/repo/https---www.phpbb.com-customise-db-composer-packages-extension.json/packages.json into cache
Downloading https://www.phpbb.com/customise/db/composer/packages-extension.json/packages-extension-1.json
Downloading https://www.phpbb.com/customise/db/composer/packages-extension.json/packages-extension-1.json
Downloading https://www.phpbb.com/customise/db/composer/packages-extension.json/packages-extension-1.json
```

Nicofuma commented:
For the last error, that's because you need to add the base url (https://www.phpbb.com/customise/db/composer/) and not the complete one (https://www.phpbb.com/customise/db/composer/packages.json)

paul999 commented:
Indeed, that works :), some more issues, they happen with my Ajax Shoutbox (Which has a soft-version constraint of <= 3.2.*@dev), and a requirement to packagist (With packagist disabled)
 * At the top of the page: [phpBB Debug] PHP Warning: in file [ROOT]/phpbb/composer/io/translate_composer_trait.php on line 225: Invalid argument supplied for foreach()
 * A empty error message: 
```
Information



« Back to previous page
```
* And, as expected the actual error:
```
Your requirements could not be resolved to an installable set of packages.

  Problem 1
    - Installation request for paul999/ajaxshoutbox * -> satisfiable by paul999/ajaxshoutbox[1.0.2].
    - paul999/ajaxshoutbox 1.0.2 requires kriswallsmith/buzz ~0.13 -> no matching package found.

Potential causes:
 - A typo in the package name
 - The package is not available 
```
I guess extensions with a soft-constraint should not be in the @phpbb/titania composer list.

The suggestion for having packagist disabled won't work for a lot of extensions, because they have a require on composer/installers. (Also, packagist support is enabled by default?)

Installing a extension works, however I don't see a option to enable it afterwards (It shows as installed, which is not logged actually?) It also doesn't show up in the available extension list.
There is also no option after installation to remove it.

When setting the enable on install option this is shown in the extension list in the ACP:
```
The “dmzx/buttonroll” extension is not valid.
The requested file could not be found: ./../ext/dmzx/buttonroll/composer.json
```
When looking at the ext/ directory, nothing is written there. However, they are written into vendor-ext:
```
root@v1:/home/paul/phpbb2fa.com/public_html/dev/phpbb/phpBB/vendor-ext# ls -hal
total 24K
drwxrwxrwx  5 paul     paul     4.0K Dec 16 10:18 .
drwxr-xr-x 20 paul     paul     4.0K Dec 16 08:53 ..
-rw-r--r--  1 www-data www-data  183 Dec 16 10:18 autoload.php
drwxr-xr-x  2 www-data www-data 4.0K Dec 16 10:18 composer
drwxr-xr-x  3 www-data www-data 4.0K Dec 16 10:18 dmzx
drwxr-xr-x  3 www-data www-data 4.0K Dec 16 10:14 hifikabin

```

Nicofuma commented:
About the settings, it may be due to a previous install.
For the other errors... the ext shouldn't be installed in vendor-ext/ but as said earlier the extension installed from titania can't be installed properly for now (it's a packaging issue).

For the "soft-version constraint" that's a concept that doesn't exist for now. That's why titania try to convert it to a real constraint IIRC. By the way your constraint is wrong: `<= 3.2.*@dev` doesn't make any sense. It should be `<= 3.2@dev` (`<=` creates a range but `3.2.*` is already a range).

About packagist... it won't work for some extension yes but `composer/installers` shouldn't be an issue (should be provided by the core but I don't remember if it's already provided, I'll check later) and for the other requirements that's something to do with titania: we need to find a way to proxify the packagist libs used by the validated extensions and not provided by the core.

> There is also no option after installation to remove it.
There should be one (in the ext list or in the catalog don't remember). But maybe it doesn't shows up because it's not in the ext/ folder

marc1706 commented:
It seems like the titania link isn't added to the extension catalog settings by default. I'd try to prevent users from having to enter the link and I'm pretty sure the link won't change in the next years. Therefore I'd hardcode the link to titania and just let users add more links to repos if they like.
The possibilites of users messing up their board by deleting the link and not knowing what the link they have to add back is worse than having this one fixed link.

marc1706 commented:
I had an issue previously where no install worked. This was due to me using PHP 7.1.0-dev. Unfortunately, the default verbosity doesn't show the actual problem the installation is encountering:
```
Loading remote repositories with package information
Updating packages

  Problem 1
    - This package requires php >=5.4,<7.1 but your PHP version (7.1.0-dev) does not satisfy that requirement.
```
(The part with Problem 1 ... only started showing up once I increased the verbosity manually in the files)

Nicofuma commented:
For the titania link it should be added by default by the migration and I don't think we should hardcode it in any other way (we should always be able to remove titania from the list)

DavidIQ commented:
Maybe we should set up a rewrite rule on .com for that.  And yes I agree that Titania should be included by default, but I don't agree it should be removable.  The later is, of course, a matter of opinion.

Nicofuma commented:
Titania is already included by default, and I truly think it should be removable. Why? because titania is a phpbb.com thing. Currently you are able to install phpBB without depending on phpb.com and it should remain like that.

Marc: solved your issue with the error message

bantu commented:
phpBB.com MUST BE removable.

marc1706 commented:
Is it being added by default now, though? Cause until now the .com repo didn't exist in the default install and I always had to add it manually.

DavidIQ commented:
I don't see what the insistence is on it being "MUST BE removable".  Unless phpbb.com falls off the face of the Internet I don't really see many users changing it anyways.

In any case as long as there is somewhere that this information can be obtained from EASILY in the case of someone wanting to put it back, which will happen, then it shouldn't be a problem for support I think.  Oh and keeping it in doesn't make it dependent at all on phpbb.com so that's an incorrect statement.

marc1706 commented:
Also, what about using a switch for having .com in there instead of letting users modify the link to it? I expect a few to remove those links by accident and then not finding the link again.

Un1matr1x commented:
If you add a new repositorie check if titania exists, if not prefill the matching field wit titania settings, so it's easy to re-add, if removed?!

marc1706 commented:
I do get this if the composer-ext.json does not exist:
```
[phpBB Debug] PHP Warning: in file [ROOT]/vendor/composer/composer/src/Composer/Json/JsonFile.php on line 86: file_get_contents(composer-ext.json): failed to open stream: No such file or directory
```
We should either make sure we always have a composer-ext.json or we should try to check if the file exists first.

marc1706 commented:
Also, is there a way we could try to prevent this from happening?
```
Fatal error: Maximum execution time of 30 seconds exceeded in phpBB\vendor\composer\composer\src\Composer\Util\RemoteFilesystem.php on line 333
```

Nicofuma commented:
That's weird, it shouldn't happen (unless you installed 1xx extensions?)

Le ven. 25 déc. 2015 08:55, Marc Alexander <notifications@github.com> a
écrit :

> Also, is there a way we could try to prevent this from happening?
>
> Fatal error: Maximum execution time of 30 seconds exceeded in phpBB\vendor\composer\composer\src\Composer\Util\RemoteFilesystem.php on line 333
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phpbb/phpbb/pull/3909#issuecomment-167207497>.
>

marc1706 commented:
Getting this right now when having packagist enabled:
```Could not load package alex75/noquote in https://www.phpbb.com/customise/db/composer: [UnexpectedValueException] Could not parse version constraint >=3.1.x: Invalid version string "3.1.x"```

Nicofuma commented:
That's because >=3.1.x is not a valid version constraint. It should be
>=3.1.0 instead (3.1.x imply a range such as >=)

Le ven. 25 déc. 2015 10:21, Marc Alexander <notifications@github.com> a
écrit :

> Getting this right now when having packagist enabled:
> Could not load package alex75/noquote in
> https://www.phpbb.com/customise/db/composer: [UnexpectedValueException]
> Could not parse version constraint >=3.1.x: Invalid version string "3.1.x"
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/phpbb/phpbb/pull/3909#issuecomment-167213369>.
>

marc1706 commented:
I have almost no extension installed. XAMPP on windows is just incredibly slow which is probably why I hit the execution limit. And of course ```>=3.1.x``` is not a valid version constraint. But that is just one extension that exists. I'm not trying to install it. This does however show up while trying to install any other extension that was retrieved via packagist.

Also, the install log is IMHO not even close to being verbose enough. It always just shows something like ```Loading remote repositories with package information``` and then some error message on top. For the version constraint issue it just shows the error message for that and nothing about the actual install.

prototech commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -129,4 +148,30 @@
+	'EXTENSIONS_REMOVED'			=> 'Extensions successfully removed.',
+	'EXTENSIONS_UPDATED'			=> 'Extensions successfully updated.',
+
+	'EXTENSIONS_COMPOSER_NOT_WRITABLE'	=> 'In order to use the catalog, the following files and directories must be writable: ext/ vendor-ext/ composer-ext.json and composer-ext.json.lock',
```
Shouldn't `composer-ext.json.lock` be `composer-ext.lock`?

Nicofuma commented:
For the log output the major issue is translation... but feel free to complete it :)

Your issue with the list should be fixed now.

Nicofuma commented:
Does someone have a summary of the remaining known issues?

marc1706 commented:
I'll try to take a look at it again today. Btw, I'd rather have log output in English than none at all.

paul999 commented:
I agree with @marc1706, beter have something good in English as nothing at all.

marc1706 commented:
A small list of current issues (feel free to edit and add your own entry):

* composer-ext.json, composer-ext.lock, and vendor-ext do not exist by default
* Missing lang entries

Nicofuma commented:
A small list of current issues (feel free to edit and add your own entry):

*  composer-ext.json, composer-ext.lock, and vendor-ext do not exist by default
* Missing lang entries
* This shouldn't appear ```[phpBB Debug] PHP Warning: in file [ROOT]/vendor/composer/composer/src/Composer/Json/JsonFile.php on line 86: file_get_contents(composer-ext.json): failed to open stream: No such file or directory```
* Titania seems to be broken right now. No packages appear in the extensions catalog right now
* :white_check_mark: We should make sure one broken package does not break the install for every other package, i.e. ```Could not load package alex75/noquote in https://www.phpbb.com/customise/db/composer: [UnexpectedValueException] Could not parse version constraint >=3.1.x: Invalid version string "3.1.x"```

Nicofuma commented:
> Titania seems to be broken right now. No packages appear in the extensions catalog right now
That's because the extension `alex75/noquote` (https://www.phpbb.com/customise/db/extension/hide_quote_button_in_last_post_2/) has been validated with an invalid composer.json file (phpbb/phpbb: >=3.1.x is present in the require section)

marc1706 commented:
You currently have merge conflicts btw.

Nicofuma commented:
updated

marc1706 commented:
You should change the permissions of composer-ext.json and composer-ext.lock to 644 instead of 755

marc1706 commented:
Just encountered this: https://gist.github.com/marc1706/86cb1f0f0459e56759ae

Nicofuma commented:
updated

Nicofuma commented:
updated

@VSEphpbb could you review the wording please?

There is currently an issue with titania, at least 3 extensions needs to be repacked

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -153,10 +153,11 @@
 	'ENABLE_PACKAGIST'				=> 'Enable Packagist usage',
-	'ENABLE_PACKAGIST_EXPLAIN'		=> 'TODO: text explaining why packagist shouldn\'t be enabled, dev usage, etc...',
-	'ENABLE_PACKAGIST_CONFIRM'		=> 'TODO: are you sure you ant to enable packagist?',
+	'ENABLE_PACKAGIST_EXPLAIN'		=> 'Allow the usage of packagist while looking for available extensions. It must be used carefully as it will add not validated extensions to the catalog.',
```
> Use packagist to search for extensions? Beware that packagist may contain extensions not validated by phpBB.

Nicofuma commented:
rebased. But it should be properly tested again

michaelcullum commented:
I'll have a look at this (inc. testing tomorrow). :)

Nicofuma commented:
rebased again

Nicofuma commented:
updated

Nicofuma commented:
And rebased again, didn't test it yet... so it should be broken

marc1706 commented:
This commit by itself will already cause a build failure due to the commit message: https://github.com/phpbb/phpbb/pull/3909/commits/b4e635246346ffa2276f7aebbd79835bdc766efc

marc1706 commented:
Closed and re-opened to cause new test builds.

Nicofuma commented:
The discussion with naderman about the performances (see the first comments) let me thought about something: in this PR, packagist is disabled by default but some extensions requires it. So it could be a good idea to do something about it (reference the validated libraries in titania repo or add a note to those extensions.. don't know yet).
Anyway, I'd prefer to keep it for later and merge this PR as is.

marc1706 commented on phpBB/adm/style/acp_ext_catalog.html:
```diff
@@ -0,0 +1,130 @@
+<p>{{ lang( 'EXTENSIONS_CATALOG_EXPLAIN') }}</p>
+
+<fieldset class="quick quick-left">
+	<span class="small"><a href="https://www.phpbb.com/go/customise/extensions/3.1" target="_blank">{{ lang('BROWSE_EXTENSIONS_DATABASE') }}</a> &bull; <a href="javascript:phpbb.toggleDisplay('catalog_settings');">{{ lang('SETTINGS') }}</a></span>
```
Shouldn't this be changed to 3.2?

marc1706 replied:
I think we should also work a bit on stuff like inline CSS and inline JS though that can and should IMHO happen after this is merged.

Nicofuma replied:
Actually I think it should be dynamic (a PHPBB_BRANCH constant maybe). But lets keep that for later

Nicofuma replied:
I agree

Nicofuma replied:
done

marc1706 commented on phpBB/adm/style/acp_ext_list.html:
```diff
@@ -89,6 +89,12 @@
+
+				<!-- IF disabled.META_NAME in MANAGED_EXTENSIONS -->
+					<!-- IF disabled.actions is defined -->&nbsp;|&nbsp;<!-- ENDIF -->
+					<a href="{U_CATALOG_ACTION}&amp;action=update&amp;extension={disabled.META_NAME}">{L_UPDATE}</a>
```
Also don't think we should be building URLs in here?

Nicofuma replied:
done

marc1706 commented on phpBB/config/default/container/services_extensions.yml:
```diff
@@ -0,0 +1,47 @@
+            - '@cache.driver'
+            - '@ext.manager'
+            - '@filesystem'
+            - phpbb-extension
```
Should we have strings like that in parameters?

Nicofuma replied:
well I don't think it is an issue as they are constant and cannot change (phpbb-extension is composer dependent)

marc1706 replied:
ok, so those are the fixed keywords? in that case I guess ok.

Nicofuma replied:
yes, they are

marc1706 commented on phpBB/config/default/container/services_extensions.yml:
```diff
@@ -0,0 +1,47 @@
+        arguments:
+            - '@ext.composer.installer'
+            - '@cache.driver'
+            - phpbb-style
```
Same as above

marc1706 commented on phpBB/config/default/container/services_extensions.yml:
```diff
@@ -0,0 +1,47 @@
+        arguments:
+            - '@ext.composer.installer'
+            - '@cache.driver'
+            - phpbb-language
```
And again

marc1706 commented on phpBB/config/development/config.yml:
```diff
@@ -11,3 +11,7 @@ core:
         enable_debug_extension: true
+
+    extensions:
+        composer_debug: true
```
I guess these default to false then for the non-dev environments?

Nicofuma replied:
yes

marc1706 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -383,6 +417,336 @@ function main()
 		$this->tpl_name = $tpl_name;
 	}
 
+	public function catalog_mode($id, $mode)
```
Docblock? :)

Nicofuma replied:
done

marc1706 commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,704 @@
+		{
+			$repositories[] = [
+				'type'	=> 'composer',
+				'url'	=> $repository,
```
I'd recommend checking that this is actually a URL.

Nicofuma replied:
done

marc1706 commented on phpBB/phpbb/composer/manager.php:
```diff
@@ -0,0 +1,320 @@
+		return $this->installer->check_requirements();
+	}
+
+	protected function normalize_version(array $packages)
```
Docblock?

Nicofuma replied:
done

Nicofuma commented:
and rebased

marc1706 commented:
I think there are still a few minor issues with the CSS (see travis). Also, I think we should try to use JS to lazy load the list of available extensions but that's something we can do after merging this.
The extension list currently does not display any extensions for me (unless using packagist). Not sure if this is still expected due to titania or if this might be because of the development version being 3.3.0-a1.

Nicofuma commented:
rebased and fixed

VSEphpbb commented on phpBB/phpbb/db/migration/data/v320/extensions_composer.php:
```diff
@@ -0,0 +1,39 @@
+*
+*/
+
+namespace phpbb\db\migration\data\v320;
```
perhaps this should be moved to a `v330` directory now?

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
 	'META_FIELD_INVALID'	=> 'Meta field %s is invalid.',
+
+	'EXTENSIONS_CATALOG_SETTINGS'	=> 'Extensions catalog settings',
+	'ENABLE_ON_INSTALL'				=> 'Enable extensions when installing',
```
change when to while here and below

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
+	'ENABLE_PACKAGIST_CONFIRM'		=> 'Are you sure you want to enable packagist?',
+	'COMPOSER_REPOSITORIES'			=> 'Repositories',
+	'COMPOSER_REPOSITORIES_EXPLAIN'	=> 'List of composer repositories (must be the base url of the packages.json file), one per line.',
+	'NO_EXTENSION_AVAILABLE'		=> 'There is no extension available for your board',
```
There are no extensions...

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
+	'EXTENSIONS_CATALOG_SETTINGS'	=> 'Extensions catalog settings',
+	'ENABLE_ON_INSTALL'				=> 'Enable extensions when installing',
+	'PURGE_ON_REMOVE'				=> 'Purge extensions when removing',
+	'ENABLE_PACKAGIST'				=> 'Enable Packagist usage',
```
"Search Packagist"

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
+	'ENABLE_ON_INSTALL'				=> 'Enable extensions when installing',
+	'PURGE_ON_REMOVE'				=> 'Purge extensions when removing',
+	'ENABLE_PACKAGIST'				=> 'Enable Packagist usage',
+	'ENABLE_PACKAGIST_EXPLAIN'		=> 'Use packagist to browse extensions? Beware that packagist may contain extensions not validated by phpBB.',
```
'Search packagist for phpBB extensions. Beware that packagist may contain extensions not validated by the phpBB validations team.'

DavidIQ replied:
"phpBB Extension Customisations Team" would probably be more appropriate. I don't think we have anything that references a "validations team" anywhere on phpbb.com.

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
+	'PURGE_ON_REMOVE'				=> 'Purge extensions when removing',
+	'ENABLE_PACKAGIST'				=> 'Enable Packagist usage',
+	'ENABLE_PACKAGIST_EXPLAIN'		=> 'Use packagist to browse extensions? Beware that packagist may contain extensions not validated by phpBB.',
+	'ENABLE_PACKAGIST_CONFIRM'		=> 'Are you sure you want to enable packagist?',
```
change 'enable' to 'search'

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
+	'ENABLE_PACKAGIST_EXPLAIN'		=> 'Use packagist to browse extensions? Beware that packagist may contain extensions not validated by phpBB.',
+	'ENABLE_PACKAGIST_CONFIRM'		=> 'Are you sure you want to enable packagist?',
+	'COMPOSER_REPOSITORIES'			=> 'Repositories',
+	'COMPOSER_REPOSITORIES_EXPLAIN'	=> 'List of composer repositories (must be the base url of the packages.json file), one per line.',
```
"Add URLs to Composer repositories of phpBB extensions to search here, one per line (must be the base url of the packages.json file)."

VSEphpbb commented on phpBB/language/en/acp/extensions.php:
```diff
@@ -130,4 +150,32 @@
+	'COMPOSER_REPOSITORIES_EXPLAIN'	=> 'List of composer repositories (must be the base url of the packages.json file), one per line.',
+	'NO_EXTENSION_AVAILABLE'		=> 'There is no extension available for your board',
+
+	'EXTENSION_MANAGED_SUCCESS'		=> 'The extension %s is now managed automatically.',
```
...is now being managed..."

VSEphpbb commented:
I too see nothing. I turned on Packagist to see some extensions, but no matter what I try to install, it only just wants to install an extension called "alex75/noquote"

> COMPOSER_CANNOT_INSTALL
Could not load package alex75/noquote in https://www.phpbb.com/customise/db/composer: [UnexpectedValueException] Could not parse version constraint >=3.1.x: Invalid version string "3.1.x"

Nicofuma commented:
That's because of titania, there is still invalid composer.json files...

paul999 commented:
@Nicofuma Do you know what extension? I guess we can fix it with submitting a internal update.

paul999 commented on phpBB/adm/style/acp_ext_catalog.html:
```diff
@@ -0,0 +1,130 @@
+<p>{{ lang( 'EXTENSIONS_CATALOG_EXPLAIN') }}</p>
+
+<fieldset class="quick quick-left">
+	<span class="small"><a href="https://www.phpbb.com/go/customise/extensions/3.2" target="_blank">{{ lang('BROWSE_EXTENSIONS_DATABASE') }}</a> &bull; <a href="javascript:phpbb.toggleDisplay('catalog_settings');">{{ lang('SETTINGS') }}</a></span>
```
Update the link to 3.3

DavidIQ replied:
Shouldn't this just be more dynamic in nature, maybe using the version constant? Otherwise you're going to end up having to update this for every major release.

Nicofuma replied:
We will need to create the link for every version but why not

paul999 commented on phpBB/bin/phpbbcli.php:
```diff
@@ -71,6 +71,11 @@
 
 register_compatibility_globals();
 
+if (is_file($phpbb_root_path . $config['exts_composer_vendor_dir'] . '/autoload.php'))
```
Maybe silicence is_file? It can throw a warning if something went wrong.

paul999 commented on phpBB/common.php:
```diff
@@ -141,6 +141,11 @@
 
 register_compatibility_globals();
 
+if (is_file($phpbb_root_path . $config['exts_composer_vendor_dir'] . '/autoload.php'))
```
Same as above

paul999 commented on phpBB/composer-ext.json:
```diff
@@ -0,0 +1,160 @@
+        "composer/ca-bundle": "1.0.3",
+        "composer/composer": "1.2.0",
+        "phpbb/phpbb": "3.2.0-RC1",
+        "crizzo/aboutus": "*",
```
Why are these two extensions in here?

Nicofuma replied:
I forgot to not commit this file...

paul999 replied:
I guess same for the lock file? Maybe it should be gitignored in that case? :P

Nicofuma replied:
yep

paul999 commented on phpBB/download/file.php:
```diff
@@ -99,6 +99,11 @@
 	/* @var $phpbb_avatar_manager \phpbb\avatar\manager */
 	$phpbb_avatar_manager = $phpbb_container->get('avatar.manager');
 
+	if (is_file($phpbb_root_path . $config['exts_composer_vendor_dir'] . '/autoload.php'))
```
same as above with is_file.

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -25,22 +25,31 @@
 class acp_extensions
 {
 	var $u_action;
```
Make this php5 style as well?

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -382,11 +422,350 @@ function main()
+
+				try
+				{
+					$this->request->enable_super_globals();
```
Add a comment why you enable super globals?

Nicofuma replied:
Actually it is a requirement or composer and I'm wondering if it should be moved to composer/installer

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -382,11 +422,350 @@ function main()
+				}
+				catch (\phpbb\exception\runtime_exception $e)
+				{
+					$this->display_composer_exception($language, $e, $composer_io);
```
Disable the super globals in the catch as well.

DavidIQ replied:
Maybe move it to a ```finally``` block?

Nicofuma replied:
I'll do that, now it is possible. But before.... 5.4...

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -382,11 +422,350 @@ function main()
+
+				try
+				{
+					$this->request->enable_super_globals();
```
Same as above (Also for the catch)

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -382,11 +422,350 @@ function main()
+
+				try
+				{
+					$this->request->enable_super_globals();
```
Same as above.

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -382,11 +422,350 @@ function main()
+
+				try
+				{
+					$this->request->enable_super_globals();
```
Same as above.

paul999 commented on phpBB/includes/acp/acp_extensions.php:
```diff
@@ -382,11 +422,350 @@ function main()
+				$start = $this->request->variable('start', 0);
+				$base_url = $this->u_action;
+
+				$this->request->enable_super_globals();
```
Why do you need to enable super globals here?

paul999 commented on phpBB/phpbb/composer/io/html_output_formatter.php:
```diff
@@ -0,0 +1,89 @@
+		= [
+			1 => 'bold',
+			4 => 'underscore',
+			//5 => 'blink',
```
Remove commented code? (Or do support blink :P)

paul999 commented on phpBB/phpbb/di/container_builder.php:
```diff
@@ -415,6 +415,12 @@ protected function load_extensions()
 			$ext_container->compile();
 
+			$config = $ext_container->get('config');
+			if (is_file($this->phpbb_root_path . $config['exts_composer_vendor_dir'] . '/autoload.php'))
```
Same as above with is_file

VSEphpbb commented:
Just ran a successful test, installing an extension. However it installed other extensions for some reason too? I tried to install board rules and it added 2 other random extensions. Here's the output:
```
Updating packages
Resolving dependencies through SAT
Dependency resolution completed in 0.006 seconds
Analyzed 169 packages to resolve dependencies
Analyzed 115 rules to resolve dependencies
  - Installing phpbb/autogroups (1.0.1)
Reading /phpBB/store/composer/cache/files/phpbb/autogroups/97e8626f9c11fff6b5d6d42b41f08e14a5373439.zip from cache
  - Installing phpbb/boardrules (1.0.2)
Reading /phpBB/store/composer/cache/files/phpbb/boardrules/d3ddf9a32bb06cc07bad5b68c2e4690422c94814.zip from cache
  - Installing crizzo/aboutus (2.0.4)
Reading /phpBB/store/composer/cache/files/crizzo/aboutus/f35b081d1b10042821d10cba1c5a47335e5e52e8.zip from cache
Generating optimized autoload files
```

VSEphpbb commented:
Also the list of extensions is not totally correct. For example, in https://www.phpbb.com/customise/db/composer/packages-extension-1.json

You can see it has phpbb/boardrules versions all the way up to v 2.0.0, but in the list it only shows an old v 1.0.2 available.

VSEphpbb commented on phpBB/phpbb/composer/manager.php:
```diff
@@ -0,0 +1,332 @@
+		{
+			$this->available_packages = $this->cache->get('_composer_' . $this->package_type . '_available');
+
+			if (true || !$this->available_packages)
```
`true` ??

paul999 commented:
@VSEphpbb  did discover another issue, if a extension was submitted before we modified titania to include phpbb/phpbb in require, it can be installed in all cases, as the version is not validated.

We need or to fix _all_ extensions that don't have the phpbb/phpbb requirement (which will be a major pain, and maybe not even possible, as we can't replace old revisions :/), or we to disallow it to install a extension if the phpbb/phpbb requirement isn't there.

DavidIQ commented:
I would go with disallowing the installation of extensions that don't have the phpbb/phpbb requirement.

Nicofuma commented:
That's a titania issue, not a core one

VSEphpbb commented:
Well it's a core issue if you want to make it possible to disallow ext's with no require:phpbb/phpbb

https://github.com/phpbb/phpbb/pull/3909/files#diff-da1ff5772f8915b4012565e66483f6d6R466

Nicofuma commented:
But why would you do that? this line is just removing versions known to be incompatible

VSEphpbb commented:
Because as mentioned above, old versions of extension from before there was a phpbb requirement are still being included. So we would like to simply ignore exts that dont have phpb/phpbb as a require. So in that code above you could skip exts that don't find the requirement ;)

Nicofuma commented:
I don't think we want to force exts to have a phpbb/phpbb requirement in core because we don't need it. But I think it is a good idea for phpbb.com as we could consider it as being an high quality source

DavidIQ commented:
I'll clean up the .com database. These are stored in a specific table column so I'll just clear out any entries that are missing the requirement.

rmcgirr83 commented:
Late to the party but

- Will it be possible to have ext authors display their own errors of why the extension can't be installed via composer?  https://www.phpbb.com/community/viewtopic.php?p=14683741#p14683741 and https://www.phpbb.com/community/viewtopic.php?p=14718921#p14718921

Nicofuma commented:
This is unrelated to this PR, but maybe we can try to work something out.

Nicofuma commented:
@paul999 @VSEphpbb every comments fixed

VSEphpbb commented:
But Titania now automatically adds the phpbb require to every validated extension, so all current valid extensions have it...It's the old revisions that pose a problem with this now.

As mentioned above, because my test board is 3.3, Board Rules current versions are not installable because they have a <3.3 constraint... But the really old revisions from 2013-2015 are being recommended by the composer installer here because they don't have any version constraints.

So basically this composer installer is trying to tell me to install Board Rules 1.0.2 which has been dead for a long time (and 2.0.0 is the most recent).

This is a problem for every extension that existed before 2016, when Titania began enforcing the phpbb requirement on new validations. This is why @DavidIQ  and @paul999 and I are asking for this to be addressed by the composer installer in some way...or it's a big problem.

Nicofuma commented:
And again, this must be addressed by titania (the provider) which has full control of its own database. And if I'm not mistaken, @DavidIQ said he was going to fix titania database

VSEphpbb commented:
Database has nothing to do with what I'm talking about. It's literally just adding an `else {continue}` here (https://github.com/phpbb/phpbb/pull/3909/files#diff-da1ff5772f8915b4012565e66483f6d6R447) to skip over any ext that is missing its require:"phpbb/phpbb" is all that's probably needed.

Nicofuma commented:
For phpBB, an extension without phpbb/phpbb requirement is perfectly valid and so should be included in the list. But phpbb.com as a stronger quality standard and shouldn't includes any extension that does not match these standards (ie: extensions without phpbb/phpbb requirement)

VSEphpbb commented:
It was discussed a while back that we would make r`equire:phpbb/phpbb` a requirement for all extension's composer files, specifically because of this composer installation process. Pretty sure @michaelcullum has that dscussion in a private gist somewhere.

marc1706 commented on phpBB/phpbb/composer/installer.php:
```diff
@@ -0,0 +1,709 @@
+		{
+			$this->restore_cwd();
+
+			if ($super_globals) {
```
Brace on new line pls.

marc1706 commented:
Now remember kids. If something's broken, it's Nicofuma's fault.

![](https://media0.giphy.com/media/WANXTFeYwLfA4/giphy.gif)

Nicofuma commented:
lol
